### PR TITLE
Madison/jax

### PR DESCRIPTION
**Pull request by madisonmay**

### CODE REVIEW

**Comment by madisonmay** [comment]
Will add some comments in the morning to make this more legible.

File: enso/utils.py [hunk 1]
```diff
@@ -4,8 +4,46 @@
 from time import strptime
 import pandas as pd
 import numpy as np
-from sklearn.model_selection import StratifiedKFold, StratifiedShuffleSplit
+import random
+from sklearn.model_selection import StratifiedKFold, StratifiedShuffleSplit, train_test_split
 from enso.config import RESULTS_DIRECTORY, FEATURES_DIRECTORY
+from collections import defaultdict
+
+class SafeStratifiedShuffleSplit(StratifiedShuffleSplit):
```
Comment (madisonmay) [inline enso/utils.py]: This class exists just to make sure that at least one example from each class is put in each split.  StratifiedShuffleSplit doesn't guarantee that by default and it's causing issues at low data volumes

File: enso/experiment/tiny_net.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
+import haiku as hk
+import jax.numpy as jnp
+import jax
+import numpy as np
+from sklearn.preprocessing import LabelBinarizer
+
+class BetterLabelBinarizer():
+
```
Comment (madisonmay) [inline enso/experiment/tiny_net.py]: LabelBinarizer changes behavior when n_classes=2 and kind of arbitrarily assigns one class to be positive and one to be negative.  This class changes that behavior to output a shape (1, 2) one hot label per example at n_classes=2.

File: enso/experiment/rationalized.py [hunk 3]
```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
...
+            length_mask: np.ndarray,
+        ):
+            batch_size = inputs.shape[0]
+
+            rationale_log_probs, clf_log_probs = model.apply(params, x=inputs, length_mask=length_mask)
+            rationale_xent = -jnp.mean(rationale_targets * rationale_log_probs * jnp.expand_dims(length_mask, -1))
+            clf_xent = -jnp.mean(targets * clf_log_probs)
+            return rationale_xent + clf_xent
+
+        @jax.jit
```
Comment (madisonmay) [inline enso/experiment/rationalized.py]: There is definitely a better way to do this that doesn't involve inlined functions but jax has some awk requirements that make this hard.

File: enso/experiment/rationalized.py [hunk 4]
```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
...
+        stacked_targets = np.vstack(rationale_targets)
+
+        rationale_proto = np.einsum('sh,st->ht', stacked_docs, stacked_targets)
+        rationale_proto /= np.einsum('st->t', stacked_targets)
+
+        return doc_vectors, rationale_targets, targets, rationale_proto
+        
+      
+    def fit(self, X, Y):
+        # No rationale means everything is a rationale
```
Comment (madisonmay) [inline enso/experiment/rationalized.py]: Need to CV over batch_size, n_epochs, probably an L1/L2 weight penalty.  Maybe need early stopping or similar but not sure because it would probably have to be on train set (not enough data to split out to val).

File: enso/experiment/rationalized.py [hunk 5]
```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
...
+                doc_tensor[i, :doc_vector.shape[0], :] = doc_vector
+            
+            length_mask = np.zeros((bsize, max_length))
+            for i in range(bsize):
+                length_mask[i, :lengths[i]] = 1.
+
+            yield (doc_tensor, length_mask)
+
+    def featurize_x(self, X):
+        doc_vectors = []
```
Comment (madisonmay) [inline enso/experiment/rationalized.py]: I should reuse code between this and featurize_x_y

File: enso/experiment/rationalized.py [hunk 6]
```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
...
+        self.target_encoder = BetterLabelBinarizer()
+
+    def _token_in_rationales(self, token, rationales):
+        for rationale in rationales:
+            if (rationale['start'] <= token.idx) and ((token.idx + len(token.text)) <= rationale['end']):
+                return True
+        return False
+
+    def train_iter(self, doc_vectors, rationale_targets, targets, batch_size=2, n_epochs=100):
+        for _ in range(n_epochs):
```
Comment (madisonmay) [inline enso/experiment/rationalized.py]: Should re-use some code between this and predict_iter.

File: enso/experiment/tiny_net.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    def __call__(self, x: jnp.ndarray, length_mask: jnp.ndarray):
+        """
+        x: (batch, seq, n_hidden)
+        """
+        rationale_logits = hk.Conv1D(
+            output_channels=self.n_classes,
+            kernel_shape=self.kernel_size,
+            w_init=hk.initializers.Constant(self.rationale_proto)
+        )(x)
+        rationale_log_probas = jax.nn.log_sigmoid(rationale_logits)
```
Comment (madisonmay) [inline enso/experiment/tiny_net.py]: Log sigmoid because I'm producing a rationale per class.

File: enso/experiment/tiny_net.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+        rationale_probas = jnp.exp(rationale_log_probas) # (batch, seq, n_classes)
+        length_mask = jnp.expand_dims(length_mask, -1)
+        masked_rationale_probas = rationale_probas * length_mask
+        lengths = jnp.sum(length_mask, axis=1)
+        clf_embed = jnp.einsum('bsh,bsc->bhc', x, masked_rationale_probas) / np.reshape(lengths, [x.shape[0], 1, 1])  # (batch, hidden, classes)
```
Comment (madisonmay) [inline enso/experiment/tiny_net.py]: Sum reduce over sequence, weighted by each rationale's proba.

From (batch, sequence, hidden), (batch, sequence, class_probas) --> (batch, hidden, classes)

File: enso/experiment/tiny_net.py [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        flat_clf_embed = jnp.reshape(clf_embed, clf_embed.shape[:1] + (clf_embed.shape[1] * clf_embed.shape[2],))
```
Comment (madisonmay) [inline enso/experiment/tiny_net.py]: Flatten last two dims to produce overall doc representation and then project down to n_classes.

This make the final hidden projection (n_classes * hidden, n_classes).  Might want to rework this to an independent projection per class representation w/ fewer weights (hidden, n_classes)?

File: enso/experiment/rationalized.py [hunk 10]
```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
...
+        self.target_encoder.fit([y[1] for y in Y])
+        self.n_classes = len(self.target_encoder.classes_)
+
+        doc_vectors, rationale_targets, targets, proto = self.featurize_x_y(X, Y)
+        train = self.train_iter(doc_vectors, rationale_targets, targets, batch_size=2, n_epochs=500)
+
+        model_fn = lambda x, length_mask: TinyNet(kernel_size=1, n_classes=self.n_classes, rationale_proto=proto)(x, length_mask)
+        model = hk.transform(model_fn)
+        rng = jax.random.PRNGKey(0)
+        sample_data_point = next(train)
```
Comment (madisonmay) [inline enso/experiment/rationalized.py]: We need a datapoint to infer some shapes on init.  Should just do thi by passing in fake data rather than actually producing an example.

File: enso/experiment/rationalized.py [hunk 11]
```diff
@@ -306,3 +313,199 @@ def _featurize(self, doc):
...
+            rationale_targets: np.ndarray,
+            targets: np.ndarray,
+            length_mask: np.ndarray,
+        ):
+            batch_size = inputs.shape[0]
+
+            rationale_log_probs, clf_log_probs = model.apply(params, x=inputs, length_mask=length_mask)
+            rationale_xent = -jnp.mean(rationale_targets * rationale_log_probs * jnp.expand_dims(length_mask, -1))
+            clf_xent = -jnp.mean(targets * clf_log_probs)
+            return rationale_xent + clf_xent
```
Comment (madisonmay) [inline enso/experiment/rationalized.py]: Probably need some weighting on these terms. Either alpha * rationale_xent + clf_xent or alpha * rationale_xent + (1-alpha) * clf_xent.
